### ISSUE TITLE
Styling with classes (get template() called only once)

### ISSUE THREAD

**Issue by timsim00** [elix/elix#95]
tldr; what would it take to call get template() for every instance on the page?

I'm experimenting with ways to style elix components.  Here's one way:

```
<elix-button data-class="f6 link  br2 ph3 pv2 mb2 dib white dim bg-green pointer"  id="btnTachyon">Elix Tachyon Button</elix-button>

<elix-button disabled data-class="f6 link  br2 ph3 pv2 mb2 dib white button-disabled bg-gray"  id="btnTachyonDisabled">Elix Tachyon Button</elix-button>
```

I can pass classes in via dataset attributes.  Then modify the component to pick them up like this:

```
get [symbols.template]() {
    console.log('get template: ${this.id} - ${this.dataset.class}')
    return template.html'
      <style>
        @import "/assets/tachyons.min.css";
        .button-disabled {
          opacity: 0.65;
          cursor: not-allowed;
        }
      </style>

      <button class="${this.dataset.class}" id="inner" role="none">
        <slot></slot>
      </button>
    ';
  }
```

But get template() is only being called once, regardless of the number of instances on the page, so both buttons are getting the same classes.

The call chain is non-trivial:

get [template] | @ | Button.js:103
  | getTemplate | @ | ShadowTemplateMixin.js:107
  | [render] | @ | ShadowTemplateMixin.js:81
  | [render] | @ | ReactiveMixin.js:138
  | [render] | @ | WrappedStandardElement.js:333
  | [render] | @ | KeyboardMixin.js:89
  | [render] | @ | FocusVisibleMixin.js:90
  | [render] | @ | AriaRoleMixin.js:24
  | [render] | @ | Button.js:86
  | render | @ | ReactiveMixin.js:98
  | connectedCallback | @ | ReactiveMixin.js:49

Thoughts?

**Comment by JanMiksovsky** [comment]
The ShadowTemplateMixin caches component templates for performance reasons, but you can turn off that behavior. Define a property getter for [symbols.hasDynamicTemplate](https://component.kitchen/elix/symbols#symbols.hasDynamicTemplate) and have it return true:

```js
get [symbols.hasDynamicTemplate]() {
  return true;
}
```

ShadowTemplateMixin should then invoke the `symbols.template` getter each time a component class is instantiated.

If you end up applying your styling concept to multiple components, you could consider turning it into a mixin. E.g., the mixin might take all attributes of the form `data-foo-class` and apply those classes to the elements with the corresponding IDs (here, `foo`). That mixin could also define the `symbols.hasDynamicTemplate` getter above, saving you from having to redefine it in each component. That would give you something a little like a homemade version of the proposed support for [CSS parts](https://meowni.ca/posts/part-theme-explainer/).

I'd be interested to know how your styling investigations turn out. The `symbols.hasDynamicTemplate` flag is actually something of a code fossil related to earlier work. Elix no longer makes any use of it anywhere, and we've been debating whether to deprecate it.

**Comment by JanMiksovsky** [comment]
I just realized I had been so focused on your Elix question that I hadn't given much thought to the actual styling approach you're attempting here. I don't think Shadow DOM will support it.

As I understand your approach, you want to define CSS classes in the outer page, then apply those class names to elements inside the Shadow DOM of a custom element. By design, the HTML Shadow DOM boundary will prevent those classes in the outer page from being picked up by elements inside the Shadow DOM subtree. That's to prevent accidental collisions between classes defined on the page and classes defined inside the component.

Unfortunately, the Shadow DOM boundary isn't permeable enough, even when you want it to be. We won't have a native way to do that until we get something like CSS parts.

That's why Elix generally defines new component classes when it wants to create stylized variations. We're looking at various forms of helpers or syntactic sugar that we could provide to make that easier.

**Comment by timsim00** [comment]
Returning true from hasDynamicTemplate works like a charm.  To your point though, it works only because I'm importing the stylesheets (bootstrap, tachyons) from within the components' template.  (get template is actually now being called twice for each button.  Haven't researched yet.)

Thinking functional css solves a lot of the collision issues.  But still using Shadow DOM anyways.

Example is here:
https://github.com/timsim00/mark42/blob/master/components/other/uHelloWorld1v1.js

**Comment by JanMiksovsky** [comment]
Great, glad that works!

Importing the stylesheets is an interesting approach — we may look into that for some uses ourselves.

It occurs to me that another way to manipulate the class on the shadow element would be to expose an attribute/property that gets reflected to that inner element at render time:

```js
class CustomButton extends ReactiveElement {

  get defaultState() {
    return Object.assign(super.defaultState, {
      innerClass: null
    });
  }

  // This property can also be set as the `inner-class` attribute.
  get innerClass() {
    return this.state.innerClass;
  }
  set innerClass(innerClass) {
    this.setState({ innerClass });
  }

  [symbols.render](changed) {
    super[symbols.render](changed);
    if (changed.innerClass) {
      if (this.state.innerClass) {
        this.$.inner.setAttribute('class', this.state.innerClass);
      } else {
        this.$.inner.removeAttribute('class');
      }
    }
  }

  get [symbols.template]() {
    return template.html`
      <style>
        .custom {
          color: red;
        }
      </style>
      <button id="inner">
        <slot></slot>
      </button>
    `;
  }

}
```

See https://codepen.io/JanMiksovsky/pen/xNvJrO?editors=1010 for a live example.

Updating the `class` attribute of the shadow element that way would have some significant advantages:

1. The page can apply and remove classes dynamically at run time. (With the dynamic template approach, only attributes specified in the initial HTML will have any effect.)
1. The template for the component will always be the same, so it could be cached.
1. Updates to the `inner-class` attribute (or `innerClass` property) would be extremely efficient, as they'd result in a single call to `setAttribute`/`removeAttribute`.

**Comment by timsim00** [comment]
Yeah, I like that approach.  Nice work!

**Comment by timsim00** [comment]
Here's a way to import global style per component type:

```
    get defaultState() {
      // console.log('*** defaultState', this.extends)
      // https://getbootstrap.com/docs/3.4/customize/
      let innerStyle = '@import "/assets/tachyons.min.css";'
      if (this.extends === 'input') {
        innerStyle += '@import "/assets/input-bootstrap.min.css";'
      } else if (this.extends === 'button') {
        innerStyle += `
          @import "/assets/button-bootstrap.min.css";
          .button-disabled {
            opacity: 0.65;
            cursor: not-allowed;
          }
        `
      }

      return Object.assign(super.defaultState, {
        innerStyle
      });
    }
```

https://github.com/timsim00/mark42/blob/master/components/elements/elix/InnerStyleMixin.js

**Comment by JanMiksovsky** [comment]
Interesting.

It looks like `innerStyle` is somewhat different than the `innerClass` example earlier. The latter is likely to change at runtime, so is appropriate to apply in `symbols.render`, while `innerStyle` looks to be defined by the class hierarchy, and therefore unlikely to change. In that case, it'd be preferable to actually do those style definitions in `symbols.template`.

Your mixin could define a `symbols.template` getter that concatenates the additional `style` element and its `@import` onto the base class template; this [CustomButton](https://github.com/elix/elix/blob/master/demos/src/CustomButton.js) components shows that kind of concatenation.

If you do decide to do the work in `symbols.render`, I'd suggest ensuring that the `querySelector` for the existing style node doesn't inadvertently pick up some other style element that happens to be in the template. You could add an `id` attribute to the style element you create, then look for an existing style element with that ID before you call `replaceWith`.

In any event, since it sounds like the original issue has been addressed, I'd like to close this. Feel free to continue to ask questions via email at jan@component.kitchen, though.
